### ISSUE TITLE
Out-of-sample/cluster Prediction

### ISSUE THREAD

**Issue by JonnyPhillips** [nealjean/predicting-poverty#16]
I've succeeded in replicating your results (great work by the way), but I'm now trying to make predictions for consumption/assets in places outside of the original DHS/LSMS clusters, i.e. out-of-sample predictions derived from additional satellite imagery from non-DHS/LSMS locations. I can see from extract_features.py that the features are estimated for every image provided before being aggregated to the cluster level, so this should be feasible. But I'm then unsure how to use these image-specific features in the regression model produced in fig_utils.py, partly because everything is coded at the cluster level, reflecting the available level of DHS/LSMS data, and partly because I'm not familiar with the cross-validation approach. How would you advise applying the regression model to make predictions for individual images? Thanks,

**Comment by nealjean** [comment]
If you want to make predictions for individual images then you can just treat those locations as clusters with only one image - instead of averaging the image features across all the images from a DHS cluster, you can just treat the image features for that image as the feature vector for a cluster at that location.

**Comment by JonnyPhillips** [comment]
Exactly - and extract_features works well for this. But extract_features returns the conv_features with 4096 measures for each image and I guess I'm unsure how to turn this into specific consumption/asset predictions for each location. When I try and adapt the 'predict_consumption' code from Fig. 3, I'm a little unsure how much of the code ('run_cv' etc.) is needed just to make the comparison with the known cluster LSMS consumption measures for Fig. 3 and how much is needed to make predictions at each location from the trained model. The results I've been able to generate only return y_hat for the original DHS clusters and not the new single-image clusters, I'm guessing because the ancillary data from consumptions.npy and other files only have data for the original DHS clusters and so only those rows are included. 
In short, I'd be grateful for any advice on what would be the correct workflow to make these out-of-sample predictions after extracting the features? Thanks,

**Comment by nealjean** [comment]
I'll try to outline the procedure that (I think) you'll want to follow. Let's assume that you want to make consumption predictions for Uganda for locations that aren't included in the LSMS clusters.

1. Collect conv_features (n, 4096) and consumptions (n,) for Uganda. Let's call the features X_lsms and the consumptions y_lsms.

2. Use StandardScaler to fit and transform X_lsms.

3. Create a sklearn ridge regression model (linear_model.Ridge) and fit it on the LSMS data. This should look something like: ridge.fit(X_lsms, y_lsms). See line 590 of fig_utils.py for an example.

4. Use the trained ridge regression model to make predictions at new locations: y_hat_new = ridge.predict(X_new)


The cross validation code is there to make sure that we are being careful about keeping our training/validation/test sets separate and getting an accurate evaluation of the quality of our models. For your purposes, I think what you want to do is fit your model on the LSMS data and then use that fitted model to make consumption predictions at new locations. Following the steps outlined above should let you do that.

Hope that helps!

**Comment by JonnyPhillips** [comment]
This was extremely helpful guidance, thank you. I managed to execute everything and generate predictions, but have stumbled on an odd problem: The extracted features data from the satellite images at the predicted points ('X_new') repeats every 32 images (features for image 1 are identical to image 33, 2 is identical to 34 etc.). Clearly that's to do with the batch processing. Basically the code is just extracting the features for 32 images and repeating this for each batch. I've verified that different images are being input and that curr_batch varies from one batch to the next. It's at the line "net.blobs['data'].data[...] = curr_batch_1_32" that the failure occurs and the values of the data somehow become the same across batches, with later batches overwriting earlier ones I think. But with limited knowledge of the caffe.Net model I could not trace why this was occurring. The issue seemed to be related to this issue here https://github.com/BVLC/caffe/issues/4654 and https://stackoverflow.com/questions/44119176/caffe-model-gives-same-output-for-every-image
I wondered if you had encountered any similar issue or could shed any light on how net.blobs['data'] is expected to behave? Thanks again,

**Comment by JonnyPhillips** [comment]
Just to follow-up, I 'resolved' this issue by literally reloading the net model at the end of every batch to ensure a clean set of features were extracted. That's obviously not ideal, but provides accurate, intuitive and non-duplicated results. I'm not sure if this is a deeper issue or a problem with my setup, but it may be worth checking. Thanks again,

**Comment by nealjean** [comment]
We've moved on from Caffe at this point so we won't be making any changes to the existing pipeline - glad that you were able to figure it out!

**Comment by reidfalconer** [comment]
Hi @JonnyPhillips, I seem to have run into the identical problem (the model is predicting roughly constant values for all locations) in attempting to replicate the results. 

You seem to have found a solution. However, given my insufficient knowledge of caffe I'm struggling to edit the code accordingly to reload the net model after every batch to ensure that a clean set of features is extracted. Could you possibly shed some light on how to go about doing this? 

Thank you in advance, and I will be grateful for any help you can provide.

**Comment by JonnyPhillips** [comment]
Not sure if this helps, I'm very much a beginner too, but I'm copying the code I adapted and used below. Note this likely contains errors and lots of inefficiencies. My memory is hazy but I think the alteration is moving the following lines (or similar) within the loop for each batch in get_features:

``` 
net.blobs['data'].data[...] = curr_batch
net.forward()
raw_features.append(net.blobs['conv7'].data)
```

```
import numpy as np
import math
import os
from scipy.misc import imread
import caffe

os.chdir('C:/Users/Jonny/Google Drive/Consultancy Work/Kaduna/Planning/Poverty Mapping/Burke Project Github/Second_Effort')

# If using GPU, set to True
GPU = False
if GPU:
  caffe.set_mode_gpu()
  caffe.set_device(0)
else:
  caffe.set_mode_cpu()


def get_features(net, locs_file):
  '''
  Runs the forward pass of the neural net on every image.
  '''
  img_cluster_locs, num_images = get_locs_info(locs_file)
  num_batches = math.ceil(num_images / 32.0)
  raw_features = []
  batch_num = 0
  with open(locs_file, 'r') as f:
    curr_batch = []
    for line in f:
      img_path = line.split()[0]
      # reads a RGB image
      input_img = imread(img_path).astype(np.float32)
      # convert to BGR
      input_img = input_img[:, :, [2, 1, 0]]
      # convert to D,H,W
      input_img = np.transpose(input_img, [2, 0, 1])
      # subtract the mean
      mean_bgr = [103.334, 107.8797, 107.4072]
      for i in xrange(0, 3):
        input_img[i, :, :] = input_img[i, :, :] - mean_bgr[i]
      curr_batch.append(input_img)

      if len(curr_batch) == 32:
        batch_num += 1
        print("Batch %d/%d for %s" % (batch_num,
                                      num_batches, locs_file))
        curr_batch = np.asarray(curr_batch)
        net.blobs['data'].data[...] = curr_batch
        net.forward()
        raw_features.append(net.blobs['conv7'].data)
        curr_batch = []
        model_file = 'model/predicting_poverty_deploy.prototxt'
        weights_file = 'model/predicting_poverty_trained.caffemodel'
        net = caffe.Net(model_file, weights_file, caffe.TEST)

    if len(curr_batch) > 0:
      batch_num += 1
      print("Batch %d/%d for %s" % (batch_num, num_batches, locs_file))
      curr_batch = np.asarray(curr_batch)
      batch_size = curr_batch.shape[0]
      # pad end batch
      curr_batch = np.vstack((curr_batch, np.zeros((32 - batch_size, 3, 400, 400)).astype(np.float32)))
      net.blobs['data'].data[...] = curr_batch
      net.forward()
      raw_features.append(net.blobs['conv7'].data[:batch_size])
      model_file = 'model/predicting_poverty_deploy.prototxt'
      weights_file = 'model/predicting_poverty_trained.caffemodel'
      net = caffe.Net(model_file, weights_file, caffe.TEST)

  raw_features = np.vstack(raw_features)
  # average pooling
  n, f, h, w = raw_features.shape
  features = raw_features.reshape(n, f, h*w)
  features = np.mean(features, axis=2)
  return features, img_cluster_locs


def aggregate_features(features, img_cluster_locs, clusters):
  '''
  Aggregate features by cluster by taking the mean.
  Respects the cluster ordering given by lats.npy and lons.npy.
  '''
  # average the features in the same cluster
  conv_features = []
  image_counts = []
  for cluster in clusters:
    cluster_mask = [(img_cluster == cluster) for img_cluster in img_cluster_locs]
    cluster_mask = np.asarray(cluster_mask)
    image_count = np.sum(cluster_mask)
    # if count is 0, fill with a 0 feature
    if image_count == 0:
      mean_cluster_feature = np.zeros(features.shape[1])
    else:
      mean_cluster_feature = np.mean(features[cluster_mask], axis=0)
      mean_cluster_feature = np.mean(features[cluster_mask], axis=0)
    conv_features.append(mean_cluster_feature)
    image_counts.append(image_count)
  conv_features = np.asarray(conv_features)
  image_counts = np.asarray(image_counts)
  return conv_features, image_counts


def extract(net, countries, output_dir):
  '''
  Runs the forward pass of the CNN on every image and then
  aggregates the features by cluster by taking the mean.
  '''
  for country in countries:
    print("Extracting %s for %s" % (country, output_dir))
    locs_file = os.path.join(output_dir, country, 'downloaded_locs_kd.txt')

    # compute conv features for every image
    features, img_cluster_locs = get_features(net, locs_file)

    # get the master cluster ordering
    cluster_lats = np.load(os.path.join(output_dir, country, 'lats.npy'))
    cluster_lons = np.load(os.path.join(output_dir, country, 'lons.npy'))
    # bit of a hack here - cluster locations can be changed when
    # writing to a file using format string
    clusters = [(float("%f" % cluster_lats[i]), float("%f" % cluster_lons[i])) for i in xrange(cluster_lats.size)]

    # aggregate features by cluster
    conv_features, image_counts = aggregate_features(features, clusters, clusters) ###***Changed middle term to clusters to avoid decimal place problem

    conv_features_path = os.path.join(output_dir, country, 'conv_features')
    image_counts_path = os.path.join(output_dir, country, 'image_counts')
    np.save(conv_features_path, conv_features)
    np.save(image_counts_path, image_counts)


def get_locs_info(locs_file):
  '''
  Get the cluster location for each image and compute the number of
  images.
  '''
  img_cluster_locs = []
  num_images = 0
  with open(locs_file, 'r') as f:
    for line in f:
      num_images += 1
      img_path, lat, lon, cluster_lat, cluster_lon = line.split()
      cluster_loc = (float(cluster_lat), float(cluster_lon))
      img_cluster_locs.append(cluster_loc)
  return img_cluster_locs, num_images


if __name__ == '__main__':
  model_file = 'model/predicting_poverty_deploy.prototxt'
  weights_file = 'model/predicting_poverty_trained.caffemodel'
  net = caffe.Net(model_file, weights_file, caffe.TEST)

  # DHS
  print("Extracting features for DHS")
  countries = ['nigeria']
  dhs_dir = 'data/output/DHS'
  extract(net, countries, dhs_dir)
```

**Comment by reidfalconer** [comment]
@JonnyPhillips Thank you so much for the response and being willing to help. Your code snippet has all the answers! 

Simply adding:
```
model_file = 'model/predicting_poverty_deploy.prototxt'
        weights_file = 'model/predicting_poverty_trained.caffemodel'
        net = caffe.Net(model_file, weights_file, caffe.TEST)
```
within the loop for each batch ensures that a clean set of features is extracted.

Thanks again 👍

**Comment by mayerven** [comment]
How to generate "conv_features.npy" in " ../data/output/LSMS/pooled" ? I can't find related code in "extract_features.py". And Figure 4 need it. Could you help me with it? Thanks~
